### PR TITLE
Support the function COL_LENGTH()

### PR DESCRIPTION
**Pull request by roshan0708**
### Description

Implement a function COL_LENGTH(table_name, column_name) which returns the Transact-SQL length of the specified column in bytes based on its data type and length modifiers.

Input-
CREATE TABLE test_table(c1 nvarchar(40), c2 varchar(MAX), c3 char(10));
SELECT COL_LENGTH('test_table', 'c1');
SELECT COL_LENGTH('test_table', 'c2');
SELECT COL_LENGTH('test_table', 'c3');
GO

Output-
80
-1
10

Signed-off-by: Roshan Kanwar [rskanwar@amazon.com](mailto:rskanwar@amazon.com)

### Issues Resolved

BABEL-3489

### Test Scenarios Covered ###
* **Use case based -** Added test cases for columns declared using all data-types supported by BBF

* **Boundary conditions -** Added test cases for handling columns declared with `MAX` specifier

* **Arbitrary inputs -** Added test cases  where inputs can be `NULL`

* **Negative test cases -** Added

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_functions.sql
Comment (hash-16) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:1]: You should add the function in upgrade script file to fix upgrade test failures.

Comment (roshan0708) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:1]: Should I add the function in `babelfishpg_tsql--3.2.0--3.3.0.sql` or any upgrade file?

Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:1]: Yes under upgrade/babelfishpg_tsql--3.2.0--3.3.0.sql .

File: test/JDBC/input/BABEL-3489-vu-verify.sql [hunk 1]
```diff
@@ -0,0 +1,98 @@
+USE babel_3489_test_db;
```
Comment (shah-nirmit) [inline test/JDBC/input/BABEL-3489-vu-verify.sql]: Can we add dependencies on this new function ?
Example View,Procedure referencing it

File: test/JDBC/upgrade/latest/schedule [hunk 2]
```diff
@@ -86,6 +86,7 @@ BABEL-3402
 BABEL-3474
 BABEL-3478
 BABEL-3486
+BABEL-3489
```
Comment (shalinilohia50) [inline test/JDBC/upgrade/latest/schedule]: Lets rename the test file as `col_length`.

File: test/JDBC/input/col_length-vu-verify.sql [hunk 3]
```diff
@@ -0,0 +1,110 @@
+USE babel_3489_test_db;
+GO
+
```
Comment (shalinilohia50) [inline test/JDBC/input/col_length-vu-verify.sql:3]: Add some tests with regular `col_length` function call (not inside view, procedures etc).
Include negative test cases too.

File: test/JDBC/input/col_length-vu-prepare.sql [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    col_smalldatetime SMALLDATETIME,
+    col_real REAL,
+    col_float FLOAT,
+    col_time TIME,
+    col_datetime DATETIME,
+    col_datetime2 DATETIME2,
+    col_datetimeoffset DATETIMEOFFSET,
+    col_uniqueidentifier UNIQUEIDENTIFIER
+);
+GO
```
Comment (shalinilohia50) [inline test/JDBC/input/col_length-vu-prepare.sql:42]: Can you please create a table in another schema and write few tests around that?

**Review by shalinilohia50 — APPROVED** [review]

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql [hunk 5]
```diff
@@ -44,3 +44,98 @@ DROP PROCEDURE sys.babelfish_drop_deprecated_object(varchar, varchar, varchar);
 
 -- Reset search_path to not affect any subsequent scripts
 SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql]: Please move the function definition before line:45

**Review by shalinilohia50 — APPROVED** [review]

**Review by shalinilohia50 — APPROVED** [review]

**Review by hash-16 — COMMENTED** [review]
Please resolve the merge conflicts.

**Review by shalinilohia50 — APPROVED** [review]

File: test/JDBC/expected/col_length-vu-prepare.out [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+RETURNS INT AS
+BEGIN
+RETURN (SELECT COL_LENGTH('sys_column_length_test_table', -1));
+END
+GO
+
+-- Invalid table, should return NULL
+CREATE FUNCTION col_length_prepare_f15()
+RETURNS INT AS
+BEGIN
```
Comment (hash-16) [inline test/JDBC/expected/col_length-vu-prepare.out:225]: You can add few having empty text as parameters , using some cast in parameters and also using this function in some tables and as expressions.

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 7]
```diff
@@ -3926,6 +3926,101 @@ RETURNS SETOF RECORD
 AS 'babelfishpg_tsql' LANGUAGE C STABLE;
 GRANT EXECUTE ON FUNCTION sys.fn_listextendedproperty TO PUBLIC;
 
+-- Matches and returns column length of the corresponding column of the given table
+CREATE OR REPLACE FUNCTION sys.COL_LENGTH(IN object_name TEXT, IN column_name TEXT)
+RETURNS SMALLINT AS $BODY$
+    DECLARE
+        col_name TEXT;
+        object_id oid;
```
Comment (hash-16) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:3934]: From which schema are we using text data type from ?

Comment (roshan0708) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:3934]: from `pg_catalog` schema

**Review by shah-nirmit — APPROVED** [review]

**Review by jsudrik — APPROVED** [review]
Based on earlier approvals.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql [hunk 8]
```diff
@@ -302,6 +302,101 @@ ALTER FUNCTION sys.power(IN arg1 TINYINT, IN arg2 NUMERIC) STRICT;
 -- Please have this be one of the last statements executed in this upgrade script.
 DROP PROCEDURE sys.babelfish_drop_deprecated_object(varchar, varchar, varchar);
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.3.0--3.4.0.sql:303]: Can you please move this to end of the script?

**Pull request merged**